operators/src/util/statistics.rs

```
@@ -0,0 +1,723 @@
+/// parabolic formula.
+///
+/// For further details, see
+///
+/// R. Jain and I. Chlamtac, The P^2 algorithm for dynamic calculation of quantiles and
+/// histograms without storing observations, Communications of the ACM,
+/// Volume 28 (October), Number 10, 1985, p. 1076-1085.
+/// <https://www.cse.wustl.edu/~jain/papers/ftp/psqr.pdf>
+///
+#[derive(Debug)]
+pub struct PSquareHistogram<T>
+where
+    T: AsPrimitive<f64>,
+{
+    markers: Vec<Marker>,
+    bucket_count: usize,
+    sample_count: u64,
+    _phantom: PhantomData<T>,
+}
+
+impl<T> PSquareHistogram<T>
+where
+    T: AsPrimitive<f64>,
+{
+    /// Creates a new histogram with `bucket_count` buckets. The
+    /// histogram is updated with all valid values from the given
+    /// `initial_samples`.
+    /// The initial samples must at least contain (`bucket_count`+1) valid elements
+    /// (i.e., elements that do not translate to `f64::NAN`).
+    ///
+    /// # Panics
+    /// If the given `bucket_count` is 0.
+    ///
+    /// # Errors
+    /// If the `initial_samples` contain less than (`bucket_count`+1) valid elements.
+    pub fn new(
+        bucket_count: usize,
+        initial_samples: &[T],
+    ) -> Result<PSquareHistogram<T>, StatisticsError> {
+        assert!(bucket_count > 0, "bucket_count must be > 0");
+
+        let mut values = Vec::with_capacity(bucket_count + 1);
+        let mut iter = initial_samples.iter();
+        let mut sample_count = 0;
+
+        for v in &mut iter {
+            let v: f64 = v.as_();
+            if !v.is_nan() {
+                values.push(v);
+                sample_count += 1;
+            }
+            if sample_count > bucket_count {
+                break;
+            }
+        }
+
+        // We require at least 5 valid initial samples
+        if sample_count <= bucket_count {
+            return Err(StatisticsError::Initialization {
+                reason: "Insufficient valid samples.".to_owned(),
+            });
+        }
```

Also here, just fail early. You could also use the ensure!-macro from snafu for that.